File: examples/io_grpc/bazel_build_java/service/armeria/src/main/kotlin/com/geekinasuit/daggergrpc/iogrpc/example/armeria/dagger/GrpcHandlersModule.java
package com.geekinasuit.daggergrpc.iogrpc.example.armeria.dagger;

import com.geekinasuit.daggergrpc.iogrpc.example.armeria.services.HelloWorldServiceAdapter;
import com.geekinasuit.daggergrpc.iogrpc.example.armeria.services.WhateverServiceAdapter;
import dagger.Module;
import dagger.Provides;
import dagger.multibindings.IntoSet;
import io.grpc.BindableService;

@Module
abstract class GrpcHandlersModule {
  @Provides
  @IntoSet
  public static BindableService helloWorldHandler(GrpcCallScopeGraph.Supplier supplier) {
    return new HelloWorldServiceAdapter(() -> supplier.callScope().helloWorld());
  }

  @Provides
  @IntoSet
  public static BindableService whateverHandler(GrpcCallScopeGraph.Supplier supplier) {
    return new WhateverServiceAdapter(() -> supplier.callScope().whatever());
  }
}

File: examples/io_grpc/bazel_build_java/service/armeria/src/main/kotlin/com/geekinasuit/daggergrpc/iogrpc/example/armeria/dagger/GrpcHandlersModule.kt
package com.geekinasuit.daggergrpc.iogrpc.example.armeria.dagger

import com.geekinasuit.daggergrpc.iogrpc.example.armeria.HelloWorldServiceAdapter
import com.geekinasuit.daggergrpc.iogrpc.example.armeria.WhateverServiceAdapter
import dagger.Module
import dagger.Provides
import dagger.multibindings.IntoSet
import io.grpc.BindableService

@Module
object GrpcHandlersModule {
  @Provides @IntoSet fun helloWorldHandler(supplier: GrpcCallScopeGraph.Supplier): BindableService =
    HelloWorldServiceAdapter { supplier.callScope().helloWorld() }

  @Provides @IntoSet fun whateverHandler(supplier: GrpcCallScopeGraph.Supplier): BindableService =
    WhateverServiceAdapter { supplier.callScope().whatever() }
}
